backend/shared/src/main/scala/zio/logging/LogFormat.scala

```diff
@@ -0,0 +1,214 @@
+ * import zio.logging.LogFormat._
+ * timestamp.fixed(32) |-| level |-| label("message", quoted(line))
+ * }}}
+ */
+trait LogFormat[+Out] { self =>
+  private[logging] def unsafeFormat(
+    builder: Out => Unit
+  ): ZLogger[Unit]
+
+  final def +(other: LogFormat[String])(implicit ev: Out <:< String): LogFormat[String] = {
+    val self = this.map(ev)
+    LogFormat { (builder, trace, fiberId, level, line, fiberRefs, spans) =>
+      self.unsafeFormat(builder)(trace, fiberId, level, line, fiberRefs, spans)
+      other.unsafeFormat(builder)(trace, fiberId, level, line, fiberRefs, spans)
+    }
+  }
+
+  final def |-|(other: LogFormat[String])(implicit ev: Out <:< String): LogFormat[String] = {
+    val self = this.map(ev)
+    LogFormat { (builder, trace, fiberId, level, line, fiberRefs, spans) =>
+      self.unsafeFormat(builder)(trace, fiberId, level, line, fiberRefs, spans)
+      builder(" ")
+      other.unsafeFormat(builder)(trace, fiberId, level, line, fiberRefs, spans)
+    }
+  }
+
+  final def color(color: LogColor)(implicit ev: Out <:< String): LogFormat[String] = {
+    val self = this.map(ev)
+    LogFormat { (builder, trace, fiberId, level, line, fiberRefs, spans) =>
+      builder(color.ansi)
+      self.unsafeFormat(builder)(trace, fiberId, level, line, fiberRefs, spans)
+      builder(LogColor.RESET.ansi)
+    }
+  }
+
+  final def concat(other: LogFormat[String])(implicit ev: Out <:< String): LogFormat[String] =
+    this + other
+
+  final def fixed(size: Int)(implicit ev: Out <:< String): LogFormat[String] = {
+    val self = this.map(ev)
+    LogFormat { (builder, trace, fiberId, level, line, fiberRefs, spans) =>
+      val tempBuilder = new StringBuilder
+      self.unsafeFormat(tempBuilder.append(_: String))(trace, fiberId, level, line, fiberRefs, spans)
+
+      val messageSize = tempBuilder.size
+      if (messageSize < size) {
+        builder(tempBuilder.take(size).appendAll(Array.fill(size - messageSize)(' ')).toString())
+      } else {
+        builder(tempBuilder.take(size).toString())
+      }
+    }
+  }
+
+  final def highlight(implicit ev: Out <:< String): LogFormat[String] =
+    highlight(defaultHighlighter(_))
+
+  final def highlight(fn: LogLevel => LogColor)(implicit ev: Out <:< String): LogFormat[String] = {
+    val self = this.map(ev)
+    LogFormat { (builder, trace, fiberId, level, line, fiberRefs, spans) =>
+      builder(fn(level).ansi)
+      self.unsafeFormat(builder)(trace, fiberId, level, line, fiberRefs, spans)
+      builder(LogColor.RESET.ansi)
+    }
+  }
+
+  final def map[Out2](fn: Out => Out2): LogFormat[Out2] = new LogFormat[Out2] {
+    override private[logging] def unsafeFormat(builder: Out2 => Unit) =
+      self.unsafeFormat((out: Out) => builder(fn(out)))
+
+    override def toLogger: ZLogger[Out2] = self.toLogger.map(fn)
+  }
+
+  final def spaced(other: LogFormat[String])(implicit ev: Out <:< String): LogFormat[String] =
+    this |-| other
+
+  def toLogger: ZLogger[Out]
```

Comment:
Rewrite this as follows:

```scala
def toLogger[A1 >: A](lft: LogFormatType[A1]): ZLogger[A1]
```

where:

```scala
final case class LogFormatType[A](unsafeMake: LogFormatBuilder[A, A])
object LogFormatType {
  val string: LogFormatType[String] = ???
}

final case class LogFormatBuilder[-In, +Out](unsafeAppend: In => Unit, unsafeResult: () => Out)
```

Because you need 3 things to implement `toLogger` generically:

1. Make a new builder for that type.
2. Given a builder:
   1. Append something to it
   2. Build the final result

The above case classes will give you what you need to do the generic `toLogger`.

Reply:
why do you think we need `LogFormatType` ? at first glance it looks a bit redundant

Reply:
You need it because you need 1 builder per each log message, so you need a way to make the builders.

Anyway, all of this is irrelevant: See my latest comment on the whole PR. I think we should make `LogFormat` monomorphic again, specialized to `String`, and introduce a corresponding `StructuredLogFormat` that is based on ZIO Schema's `DynamicValue` (or even just `Map[String, String]`).

Reply:
This comment:

<https://github.com/zio/zio-logging/pull/363#issuecomment-956431679>

Reply:
unless you meant `final case class LogFormatType[A](unsafeMake: () => LogFormatBuilder[A, A])`

Reply:
@jdegoes I agree with just different log format that will be for json and produce string.